collector/internal/controller.go

```diff
@@ -42,18 +60,22 @@ func NewController(cfg *controller.Config, rs receiver.Settings,
 		GRPCConnectionTimeout:  intervals.GRPCConnectionTimeout(),
 		ReportInterval:         intervals.ReportInterval(),
 		SamplesPerSecond:       cfg.SamplesPerSecond,
-	}, nextConsumer)
+	}
+
+	rep, err := controllerOption.ReporterFactory(reporterConfig, nextConsumer)
 	if err != nil {
 		return nil, err
 	}
 	cfg.Reporter = rep
+	cfg.ExecutableReporter = controllerOption.ExecutableReporter
 
 	// Provide internal metrics via the collectors telemetry.
 	meter := rs.MeterProvider.Meter(ctrlName)
 	metrics.Start(meter)
 
 	return &Controller{
-		ctlr: controller.New(cfg),
+		ctlr:       controller.New(cfg),
+		onShutdown: controllerOption.OnShutdown,
```

Comment:
I see the need for this. But atomic PRs are much easier to review. Could you split this into another PR?